trio/_core/_run.py

```diff
@@ -1223,7 +1224,8 @@ class GuestState:
     run_sync_soon_not_threadsafe = attr.ib()
     done_callback = attr.ib()
     unrolled_run_gen = attr.ib()
-    unrolled_run_next_send = attr.ib(factory=lambda: Value(None))
+    _value_factory: Callable[[], Value] = lambda: Value(None)
+    unrolled_run_next_send = attr.ib(factory=_value_factory, type=Outcome)
```

Comment:
You can use `cast()` to write this without introducing an additional class-level name. (`cast(Callable[[], Value], lambda: Value(None))`). I'm also curious what breaks if you write it on one line without the cast.

Reply:
The error message without the lambda type hint is `Returning Any from function declared to return "object"`. Indeed according to `reveal_type` the type of `lambda: Value(None)` is `'def () -> Any'`.

So far so good, but if I switch to your inline cast which specifies the `Value` return type, I get `trio/_core/_run.py:1227: error: Redundant cast to "Callable[[], Any]"`! It's maybe a variance issue?

Reply:
Oh, this is probably because `outcome` is not typed, so mypy is replacing everything imported from it with `Any`.

We should add type hints to `outcome`. In the meantime, your workaround here seems fine, just note that we're not actually getting anything from declaring the type as `Outcome` as opposed to `Any`.